### PR TITLE
[Deepbook ] Enhancement

### PR DESCRIPTION
**Pull request by wiggins-dev**
## Description 

- enhance overflow check for mathematical calculations
- explain why underflow is permitted code clean in test code
- add test to cover corner case in match_bid
- add test to cover corner case in match_bid
- add test to successfully place a limit order of type POST OR ABORT
- add test to place limit order of type invalid restrictions
- add test to cover ask side in batch cancel order
- add test to get single  order status on bid side
- add test to get level2 book status when time expired

## Test Plan 

cd to deepbook dir and run "sui move test"

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-framework/packages/deepbook/sources/math.move [hunk 1]
```diff
@@ -20,11 +29,71 @@ module deepbook::math {
         (is_round_down, ((x * y / FLOAT_SCALING_U128) as u64))
     }
 
+    public fun mul_round_check_underflow(x: u64, y: u64): (bool, u64) {
```
Comment (sblackshear) [inline crates/sui-framework/packages/deepbook/sources/math.move:32]: For each usage of `clob_math::mul` or `clob_math::div_round` in clob.move, could you please add a detailed comment explaining why underflow is acceptable?

File: crates/sui-framework/packages/deepbook/tests/clob_tests.move [hunk 2]
```diff
@@ -840,6 +852,327 @@ module deepbook::clob_test {
...
+            let account_cap = test::take_from_sender<AccountCap>(&test);
+            let account_cap_user = object::id(&account_cap);
+            let (base_custodian, quote_custodian) = clob::borrow_mut_custodian(&mut pool);
+            custodian::deposit(base_custodian, mint_for_testing<SUI>(1000, ctx(&mut test)), account_cap_user);
+            custodian::deposit(quote_custodian, mint_for_testing<USD>(10000, ctx(&mut test)), account_cap_user);
+            test::return_shared(pool);
+            test::return_to_sender<AccountCap>(&test, account_cap);
+        };
+
+        // alice palce series limit order
```
Comment (sblackshear) [inline crates/sui-framework/packages/deepbook/tests/clob_tests.move:877]:
```suggestion
        // alice place series limit order
```

Comment (wiggins-dev) [reply crates/sui-framework/packages/deepbook/tests/clob_tests.move]: we did code clean in test code just now.

File: crates/sui-framework/packages/deepbook/sources/math.move (hunk 1 above)
Comment (wiggins-dev) [reply crates/sui-framework/packages/deepbook/sources/math.move:32]: We add comments for why underflow is permitted in clob.move.

Comment (sblackshear) [reply crates/sui-framework/packages/deepbook/sources/math.move:32]: Upon further reflection, I was wondering if we shouldn't switch these--e.g., `mul` is the function that checks for overflow, and the one that does not is called `unsafe_mul` or `mul_with_underflow_risk` or similar so the caller is aware of the risk. That feels safer than a comment--WDYT?

File: crates/sui-framework/packages/deepbook/sources/clob.move [hunk 3]
```diff
@@ -390,13 +390,15 @@ module deepbook::clob {
...
+                        maker_base_quantity,
+                        maker_order.price
+                    );
                     if (flag) maker_quote_quantity = maker_quote_quantity + 1;
-                    (flag, maker_quote_quantity) = clob_math::mul_round(
+                    maker_quote_quantity = clob_math::safe_mul(
                         maker_quote_quantity,
                         FLOAT_SCALING + pool.taker_fee_rate
                     );
-                    if (flag) maker_quote_quantity = maker_quote_quantity + 1;
```
Comment (sblackshear) [inline crates/sui-framework/packages/deepbook/sources/clob.move:399]: Can you please add this change (and the test that failed before the change + passes after) in a separate PR instead of in this one? We should avoid semantics-preserving refactoring and behavior changes.

**Review by sblackshear — COMMENTED** [review]
Looks like the genesis snapshot needs to be updated; `cargo test;` and `cargo insta review` in `sui-genesis`.

**Review by sblackshear — CHANGES_REQUESTED** [review]
Please squash into one commit

File: crates/sui-framework/packages/deepbook/sources/custodian.move [hunk 4]
```diff
@@ -37,6 +37,9 @@ module deepbook::custodian {
         custodian: &Custodian<Asset>,
         user: ID
     ): (u64, u64) {
+        if (!table::contains(&custodian.account_balances, user)) {
```
Comment (sblackshear) [inline crates/sui-framework/packages/deepbook/sources/custodian.move:40]: Same here--please separate into a separate PR with the test.

File: crates/sui-framework/packages/deepbook/sources/math.move [hunk 5]
```diff
@@ -6,25 +6,104 @@ module deepbook::math {
...
+    // also returns whether the result is rounded down
+    public fun safe_div_round(x: u64, y: u64): (bool, u64) {
+        let (is_round_down, result) = unsafe_div_round(x, y);
+        assert!(result > 0, EUnderflow);
+        (is_round_down, result)
+    }
+
+    #[test]
+    fun test_mul() {
+        assert!(unsafe_mul(1_000_000_000, 1) == 1, 0);
```
Comment (sblackshear) [inline crates/sui-framework/packages/deepbook/sources/math.move:64]: Import `sui::test::assert_eq` and then use

```suggestion
        assert_eq!(unsafe_mul(1_000_000_000, 1), 1);
```

here and elsewhere for better test output

File: crates/sui-framework/packages/deepbook/sources/math.move [hunk 6]
```diff
@@ -6,25 +6,104 @@ module deepbook::math {
...
+        assert!(unsafe_mul(9_999_999_999, 1) == 9, 0);
+        assert!(unsafe_mul(9_000_000_000, 1) == 9, 0);
```
Comment (sblackshear) [inline crates/sui-framework/packages/deepbook/sources/math.move:66]: These tests are pretty limited--always multiplying by 1 is not that that interesting. Please add more.

Comment (sblackshear) [reply crates/sui-framework/packages/deepbook/sources/math.move:66]: And same for the tests below.

File: crates/sui-framework/packages/deepbook/sources/custodian.move (hunk 4 above)
Comment (wiggins-dev) [reply crates/sui-framework/packages/deepbook/sources/custodian.move:40]: Good suggestion, I will do that!

File: crates/sui-framework/packages/deepbook/sources/math.move (hunk 5 above)
Comment (wiggins-dev) [reply crates/sui-framework/packages/deepbook/sources/math.move:64]: Get it!

File: crates/sui-framework/packages/deepbook/sources/math.move (hunk 6 above)
Comment (wiggins-dev) [reply crates/sui-framework/packages/deepbook/sources/math.move:66]: Here I am testing on the extreme bounds of the math operation, mainly to reach 9_999_999_999.

File: crates/sui-framework/packages/deepbook/sources/clob.move (hunk 3 above)
Comment (wiggins-dev) [reply crates/sui-framework/packages/deepbook/sources/clob.move:399]: Adopt!

Comment (wiggins-dev) [reply crates/sui-framework/packages/deepbook/sources/clob.move:399]: Looks like the genesis snapshot needs to be updated; cargo test; and cargo insta review in sui-genesis.

For this, 
My original plan was to update the snapshot after perfecting and testing all the codes that have not been covered,
And squash into one commit to prepare for merge.

File: crates/sui-framework/packages/deepbook/sources/math.move (hunk 6 above)
Comment (wiggins-dev) [reply crates/sui-framework/packages/deepbook/sources/math.move:66]: I will add more common test cases to improve the test.

Comment (wiggins-dev) [reply crates/sui-framework/packages/deepbook/sources/math.move:66]: Sorry I pushed another commit before I see your git comments, I will separate some changes of the new commits into 
a new PR then.

**Comment by wiggins-dev** [comment]
@sblackshear I picked out the commit of additional test added,  the overflow check for mathematical calculations and 
comments for why underflow is permitted code clean in test code.
Maybe we can first merge these changes first. And for those redundant round up and fix to account balance, We will put them into the next PR.

File: crates/sui-framework/packages/deepbook/sources/math.move [hunk 7]
```diff
@@ -6,25 +6,123 @@ module deepbook::math {
...
-    public fun mul(x: u64, y: u64): u64 {
-        let x = (x as u128);
-        let y = (y as u128);
-        ((x * y / FLOAT_SCALING_U128) as u64)
+    // <<<<<<<<<<<<<<<<<<<<<<<< Error codes <<<<<<<<<<<<<<<<<<<<<<<<
+    const EUnderflow: u64 = 1;
+    // <<<<<<<<<<<<<<<<<<<<<<<< Error codes <<<<<<<<<<<<<<<<<<<<<<<<
+
+    // multiply two floating numbers
+    public fun unsafe_mul(x: u64, y: u64): u64 {
```
Comment (sblackshear) [inline crates/sui-framework/packages/deepbook/sources/math.move:14]:
```suggestion
    public(friend) fun unsafe_mul(x: u64, y: u64): u64 {
```

File: crates/sui-framework/packages/deepbook/sources/math.move [hunk 8]
```diff
@@ -6,25 +6,123 @@ module deepbook::math {
...
+    #[expected_failure(abort_code = EUnderflow)]
+    fun test_div_round_check_underflow() {
+        safe_div_round(1, 1_000_000_001);
+    }
+
+    const DEPRECATED: u64 = 0;
+
+    // === Deprecated ===
+    /// Deprecated since v1.2.0, use `safe_mul` instead.
+    public fun mul(_x: u64, _y: u64): u64 {
```
Comment (sblackshear) [inline crates/sui-framework/packages/deepbook/sources/math.move:115]: Let's delete `safe_mul` and keep this to avoid bloat?

File: crates/sui-framework/packages/deepbook/sources/math.move [hunk 9]
```diff
@@ -6,25 +6,123 @@ module deepbook::math {
...
+        abort DEPRECATED
+    }
+
+    /// Deprecated since v1.2.0, use `safe_mul_round` instead.
+    public fun mul_round(_x: u64, _y: u64): (bool, u64) {
```
Comment (sblackshear) [inline crates/sui-framework/packages/deepbook/sources/math.move:120]: Same here.

File: crates/sui-framework/packages/deepbook/sources/math.move [hunk 10]
```diff
@@ -6,25 +6,123 @@ module deepbook::math {
...
+        abort DEPRECATED
+    }
+    /// Deprecated since v1.2.0, use `safe_div_round` instead.
+    public fun div_round(_x: u64, _y: u64): (bool, u64) {
```
Comment (sblackshear) [inline crates/sui-framework/packages/deepbook/sources/math.move:124]: and here.

File: crates/sui-framework/packages/deepbook/sources/math.move (hunk 7 above)
Comment (sblackshear) [reply crates/sui-framework/packages/deepbook/sources/math.move:14]: This will give us flexibility to change in the future

File: crates/sui-framework/packages/deepbook/sources/math.move [hunk 11]
```diff
@@ -6,25 +6,123 @@ module deepbook::math {
...
+    // also returns whether the result is rounded down
+    public fun safe_mul_round(x: u64, y: u64): (bool, u64) {
+        let (is_round_down, result) = unsafe_mul_round(x, y);
+        assert!(result > 0, EUnderflow);
+        (is_round_down, result)
+    }
+
+    // divide two floating numbers
+    // also returns whether the result is rounded down
+    public fun unsafe_div_round(x: u64, y: u64): (bool, u64) {
```
Comment (sblackshear) [inline crates/sui-framework/packages/deepbook/sources/math.move:46]:
```suggestion
    public(friend) fun unsafe_div_round(x: u64, y: u64): (bool, u64) {
```

**Review by sblackshear — CHANGES_REQUESTED** [review]

**Review by sblackshear — DISMISSED** [review]
Almost there!

**Comment by sblackshear** [comment]
> I picked out the commits of additional test cases, the overflow check for mathematical calculations and
comments for why underflow is permitted code clean in test code. Maybe we can first merge these changes. And for those redundant round up and fix to account balance, We will put them into the next PR.

That works for me.

File: crates/sui-framework/packages/deepbook/sources/math.move (hunk 8 above)
Comment (alexweb30) [reply crates/sui-framework/packages/deepbook/sources/math.move:115]: Sorry, could you elaborate? Do you mean by we change all the safe_mul function name to mul? and deprecate the safe_mul?

Comment (wiggins-dev) [reply crates/sui-framework/packages/deepbook/sources/math.move:115]: Yeah, I don't fully understand what you mean to express here either.

Comment (sblackshear) [reply crates/sui-framework/packages/deepbook/sources/math.move:115]: What I mean is:
- Before this PR, there was a function called `mul`. It implemented unsafe multiplication
- This PR introduces functions called `unsafe_mul` and `safe_mul` that do the expected thing
- Instead, this PR should change the body of `mul` to implement safe multiplication, and remove `unsafe_mul`

File: crates/sui-framework/packages/deepbook/sources/math.move (hunk 7 above)
Comment (wiggins-dev) [reply crates/sui-framework/packages/deepbook/sources/math.move:14]: Here, if we change the modifier from public to public(friend), it will lead to more restrictions on the math module, and the math module does not have any behavior to modify the state, so  can you give some explanations why this can give us more flexibiluity to change?

File: crates/sui-framework/packages/deepbook/sources/clob.move [hunk 12]
```diff
@@ -548,14 +559,16 @@ module deepbook::clob {
                     emit_order_canceled<BaseAsset, QuoteAsset>(pool_id, maker_order);
                 } else {
                     let filled_base_quantity =
-                        if (taker_base_quantity_remaining >= maker_base_quantity) { maker_base_quantity }
+                        if (taker_base_quantity_remaining > maker_base_quantity) { maker_base_quantity }
```
Comment (sblackshear) [inline crates/sui-framework/packages/deepbook/sources/clob.move:562]: This looks like a bugfix--can we please add it in a separate PR with a test that exercises the bug?

**Review by sblackshear — APPROVED** [review]
LGTM other than the `>` -> `>=` change--thanks!

**Review by sblackshear — APPROVED** [review]
LGTM other than the `>` -> `>=` change--thanks!

**Review by sblackshear — APPROVED** [review]
LGTM other than the `>` -> `>=` change--thanks!
